### PR TITLE
Added UniformDrawdownProfileParameter

### PR DESCRIPTION
**Pull request by jetuk**
This commit adds a new parameter that returns a recurring profile that draws down from 1.0 to 0.0 over the course of a year. The day on which to start the drawdown is configurable with the same keywords
as to `AnnualVirtualStorage`.

Basic tests included. Also added a more complex `AnnualVirtualStorage` test model that uses the new `UniformDrawdownProfileParameter` as a control curve in a dynamic cost calculation.

### CODE REVIEW

File: pywr/parameters/_parameters.pyx [hunk 1]
```diff
@@ -833,6 +833,76 @@ cdef class ScenarioDailyProfileParameter(Parameter):
 ScenarioDailyProfileParameter.register()
 
 
+cdef class UniformDrawdownProfileParameter(Parameter):
+    """Parameter which provides a uniformly reducing value from one to zero.
+
+     This parameter is intended to be used with an `AnnualVirtualStorage` node to provide a profile
+     that represents perfect average utilisation of the annual volume. It returns a value one the
+     reset day, and subsequently reduces by 1/366 every day afterward.
```
Comment (Batch21) [inline pywr/parameters/_parameters.pyx]: Should this be : 'It returns a value of 1 on the ....'

File: pywr/parameters/_parameters.pyx [hunk 2]
```diff
@@ -833,6 +833,76 @@ cdef class ScenarioDailyProfileParameter(Parameter):
...
+        # Note that this is zero-based
+        self._reset_idoy = pandas.Period(year=2016, month=self.reset_month, day=self.reset_day, freq='D').dayofyear - 1
+
+    cpdef double value(self, Timestep ts, ScenarioIndex scenario_index) except? -1:
+        cdef int current_idoy = ts.dayofyear - 1
+        cdef int total_days_in_period
+        cdef int days_into_period
+        cdef int year = ts.year
+
+        if not is_leap_year(ts.year):
```
Comment (Batch21) [inline pywr/parameters/_parameters.pyx]: use typed `year`  variable from above?

File: pywr/parameters/_parameters.pyx [hunk 3]
```diff
@@ -833,6 +833,76 @@ cdef class ScenarioDailyProfileParameter(Parameter):
...
+        if days_into_period < 0:
+            # We're not past the reset day yet; use the previous year
+            year -= 1
+
+        if self._reset_idoy > 59:
+            # Reset occurs after February therefore next year's February might be a leap year?
+            year += 1
+
+        # Determine the number of days in the period based on whether there is a leap year or not in the current period
+        if is_leap_year(year):
```
Comment (Batch21) [inline pywr/parameters/_parameters.pyx]: This is already check above. Could the contents of this conditional be shifted there?

Comment (jetuk) [reply pywr/parameters/_parameters.pyx]: `year` could be modified by this point by the intervening checks on where we are relative to the possible leap day.

Comment (Batch21) [reply pywr/parameters/_parameters.pyx]: yeah, good point!

File: pywr/parameters/_parameters.pyx (hunk 2 above)
Comment (jetuk) [reply pywr/parameters/_parameters.pyx]: `ts.year` should be a typed attribute because `ts` is typed.

Comment (jetuk) [reply pywr/parameters/_parameters.pyx]: In fact I've confirmed by looking at the Cython `--annotate` output.

File: pywr/parameters/_parameters.pyx (hunk 1 above)
Comment (jetuk) [reply pywr/parameters/_parameters.pyx]: Fixed

**Pull request merged**